### PR TITLE
Add NSG traffic analytics to Deploy-NSG-FlowLogs

### LINKED ISSUE #83: Deploy-NSG-FlowLogs does not deploy unique resources
https://github.com/Azure/Enterprise-Scale/blob/fc5f5d97f59c327a54f143b949e04afe38f0181e/azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json#L92

This line means each deployment has the same name and does not work when there are multiple NSGs in the same region under the scope of this policy.

PR incoming...

### PR DESCRIPTION
**Pull request by matt-FFFFFF**
**This PR fixes**

* fixes #83 
* Adds the ability to optionally deploy Traffic Analytics feature to flow logs using additional parameters

pinging @victorar an @krnese 

/cc @liamfoneill 

Thanks

### CODE REVIEW

File: azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json [hunk 1]
```diff
@@ -23,6 +23,41 @@
...
+              "type": "Boolean",
+              "metadata": {
+                "displayName": "Enable Traffic Analytics"
+              },
+              "defaultValue": false
+            },
+            "flowAnalyticsWorkspaceId": {
+              "type": "string",
+              "metadata": {
+                "displayName": "Workspace GUID for Log Analytics"
```
Comment (krnese) [inline azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:44]: can you change this parameter to be "logAnalytics" and use "strongType": "oms" ?
That will give customer a better UX when assigning via portal

File: azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json [hunk 2]
```diff
@@ -37,11 +72,19 @@
                 "roleDefinitionIds": [
                   "/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c"
                 ],
-                "name": "[concat('NetworkWatcher_', field('location'), '/', 'Microsoft.Network', resourceGroup().name, field('name'))]",
+                "name": "[concat('NetworkWatcher_', field('location'), '/', field('name'), '-', resourceGroup().name, '-flowlog' )]",
```
Comment (krnese) [inline azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:75]: "name" can be removed as we don't want policy to check for resource name explicitly (as a general rule). I noticed this is not added by you, but would be grateful if you could help to remove :-)

File: azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json (hunk 1 above)
Comment (matt-FFFFFF) [reply azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:44]: Sure thing - I will update the flowAnalyticsWorkspaceResourceId parameter as it is this that refers to the ARM resource.

The parameter you have highlighted is the GUID of the workspace and is a separate property.

Comment (krnese) [reply azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:44]: Got it. If it's the workspaceId (not workspaceResourceId) then you can ignore the "strongType": "oms" for the parameter, and consider using the reference function instead (wrapped into if condition). So when providing the logAnalytics resourceId, ARM will retrieve the guid during runtime

Comment (matt-FFFFFF) [reply azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:44]: I also need this parameter to be optional - does strongType allow for empty entries?

Comment (matt-FFFFFF) [reply azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:44]: Will investigate the reference function too - thanks :)

File: azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json (hunk 2 above)
Comment (matt-FFFFFF) [reply azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:75]: Understood - thanks

File: azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json (hunk 1 above)
Comment (matt-FFFFFF) [reply azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:44]: Am testing now - will report back later today

File: azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json (hunk 2 above)
Comment (matt-FFFFFF) [reply azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json:75]: Without the `name` key the policy does not find any resources in scope. I think this is caused by the flow logs object that we are actually interested in being separate from the NSG itself, in this case it's in the `NetworkWatcherRG`

File: azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json (hunk 1 above)
Comment (matt-FFFFFF) [reply azopsreference/3fc1081d-6105-4e19-b60c-1ec1252cf560/contoso/.AzState/Microsoft.Authorization_policyDefinitions-Deploy-Nsg-FlowLogs.parameters.json]: This works well - thanks for the tips

**Comment by matt-FFFFFF** [comment]
@krnese please can you have a quick look again?

* I have improved the params using strongTypes and implemented reference() to reduce the required user input
* I could not remove the name key, if I did the policy would not bring any resources in-scope - thoughts?

**Comment by matt-FFFFFF** [comment]
Marked as draft because two more issues have surfaced:

1. Max name length for flowlog object is 80 chars - implement substring to fix
2. Storage acct muct be in same region as NSG - implement cross sub deployment to the management sub to create the storage account as needed

**Comment by krnese** [comment]
Closing this as we have made the requisite changes. Thanks for calling this out and adding the LA support.

**Pull request closed without merging**